README.rst
========================
Windows Semaphore ctypes
========================


.. image:: https://img.shields.io/pypi/v/semaphore_win_ctypes.svg
        :target: https://pypi.python.org/pypi/semaphore_win_ctypes
        :alt: Latest Version

.. image:: https://readthedocs.org/projects/semaphore-win-ctypes/badge/?version=latest
        :target: https://semaphore-win-ctypes.readthedocs.io/en/latest/?version=latest
        :alt: Documentation Status


A ctypes wrapper for Windows Semaphore Objects


* Free software: MIT license
* Documentation: https://semaphore-win-ctypes.readthedocs.io.

Use Case
--------

This module was created to overcome limitations of Python's `built-in semaphore support`_.
Specifically, the built-in semaphores:

* Can only be `shared between python processes`_ if a common parent python process coordinates
* Cannot be acquired by code written in other programming languages

Instead, this module provides wrappers for these low level Windows Semaphore APIs:

* `CreateSemaphoreExW`_
* `OpenSemaphoreW`_
* `WaitForSingleObject`_
* `ReleaseSemaphore`_
* `CloseHandle`_

Since the Windows Semaphore API uses named semaphores to permit sharing between processes, you can now share your semaphore more freely.
Check the documentation of those APIs for details about how semaphores behave on Windows.

Dependencies
------------

This module requires no runtime dependencies.

See Also
--------

Documentation:

* https://semaphore-win-ctypes.readthedocs.io/

Related Python standard library code:

* `Python threading.Semaphore`_
* `Python multiprocessing.Semaphore`_
* `Python asyncio.Semaphore`_

Similar work on other platforms:

* `POSIX IPC`_ for better semaphores on POSIX (I.E. Linux, UNIX-like, etc.) OSes.

Other items:

* `windows semaphore helper`_ -- C++ demo of working with Windows Semaphore APIs

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.


.. _`shared between python processes`: https://stackoverflow.com/a/28854553
.. _`built-in semaphore support`: https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Semaphore
.. _`CreateSemaphoreExW`: https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createsemaphoreexw
.. _`OpenSemaphoreW`: https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-opensemaphorew
.. _`WaitForSingleObject`: https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject
.. _`ReleaseSemaphore`: https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-releasesemaphore
.. _`CloseHandle`: https://docs.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-closehandle
.. _`Cookiecutter`: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage
.. _`Python threading.Semaphore`: https://docs.python.org/3/library/threading.html#threading.Semaphore
.. _`Python multiprocessing.Semaphore`: https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Semaphore
.. _`Python asyncio.Semaphore`: https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore
.. _`POSIX IPC`: https://semanchuk.com/philip/posix_ipc/
.. _`windows semaphore helper`: https://github.com/ralexander-phi/windows-semaphore-helper



HISTORY.rst
=======
History
=======

0.1.2 (2021-03-14)
------------------

* Minor bug fixes

0.1.1 (2021-03-14)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for semaphore_win_ctypes."""


tests/test_multiprocess.py
import datetime
import pytest
import subprocess
import time
import uuid

from multiprocessing.pool import ThreadPool
from semaphore_win_ctypes import AcquireSemaphore, \
    CreateSemaphore, OpenSemaphore, SemaphoreWaitTimeoutException

TEST_SEMAPHORE_MAX_COUNT = 2
TEST_THREADS = TEST_SEMAPHORE_MAX_COUNT * 2
TEST_SEMAPHORE_CREATE_TIME_MS = 5000
TEST_SEMAPHORE_ACQUIRE_HOLD_TIME_S = 2
LAST_WAS_CTYPES = False


def unique_name():
    return str(uuid.uuid4())


@pytest.fixture
def ctypes_semaphore() -> str:
    """
    Create a semaphore using ctypes
    """
    name = unique_name()
    with CreateSemaphore(name=name, maximum_count=TEST_SEMAPHORE_MAX_COUNT):
        yield name


@pytest.fixture
def cpp_semaphore() -> str:
    """
    Create a semaphore using compiled C++ helper
    """
    name = unique_name()
    proc = subprocess.Popen([
        "SemaphoreHelper.exe",
        "create",
        name,
        str(TEST_SEMAPHORE_CREATE_TIME_MS),
        str(TEST_SEMAPHORE_MAX_COUNT),
        str(TEST_SEMAPHORE_MAX_COUNT)
    ])
    # let it start
    time.sleep(1)
    yield name
    proc.wait()


def mixed_acquire(params) -> int:
    name = params[0]
    index = params[1]
    if index % 2 == 0:
        c = cpp_acquire(name)
    else:
        c = ctypes_acquire(name)
    return c


def ctypes_acquire(name: str) -> int:
    try:
        with OpenSemaphore(name=name) as sem:
            with AcquireSemaphore(sem, timeout_ms=0):
                time.sleep(TEST_SEMAPHORE_ACQUIRE_HOLD_TIME_S)
        return 1
    except SemaphoreWaitTimeoutException:
        return 0


def cpp_acquire(name: str) -> int:
    p = subprocess.run([
        "SemaphoreHelper.exe",
        "open-and-acquire",
        name,
        str(TEST_SEMAPHORE_ACQUIRE_HOLD_TIME_S * 1000)
    ])
    if p.returncode == 0:
        return 1
    else:
        return 0


def test_multiprocess_ctypes(ctypes_semaphore: str):
    # Create subprocesses, let each try to acquire the Semaphore
    with ThreadPool(TEST_THREADS) as p:
        results = p.map(
            ctypes_acquire,
            [ctypes_semaphore for _ in range(TEST_THREADS)]
        )
    # Only some of them should succeed
    assert sum(results) == TEST_SEMAPHORE_MAX_COUNT


def test_multiprocess_ctypes_with_cpp_semaphore(cpp_semaphore: str):
    # Create subprocesses, let each try to acquire the Semaphore
    with ThreadPool(TEST_THREADS) as p:
        results = p.map(
            ctypes_acquire,
            [cpp_semaphore for _ in range(TEST_THREADS)]
        )
    # Only some of them should succeed
    assert sum(results) == TEST_SEMAPHORE_MAX_COUNT


def test_multiprocess_cpp_acquire_with_ctypes_semaphore(ctypes_semaphore: str):
    # Create subprocesses, let each try to acquire the Semaphore
    with ThreadPool(TEST_THREADS) as p:
        results = p.map(
            cpp_acquire,
            [ctypes_semaphore for _ in range(TEST_THREADS)]
        )
    # Only some of them should succeed
    assert sum(results) == TEST_SEMAPHORE_MAX_COUNT


def test_multiprocess_cpp_acquire_with_cpp_semaphore(cpp_semaphore: str):
    # Create subprocesses, let each try to acquire the Semaphore
    with ThreadPool(TEST_THREADS) as p:
        results = p.map(
            cpp_acquire,
            [cpp_semaphore for _ in range(TEST_THREADS)]
        )
    # Only some of them should succeed
    assert sum(results) == TEST_SEMAPHORE_MAX_COUNT


def test_multiprocess_mixed_with_cpp_semaphore(cpp_semaphore: str):
    # Create subprocesses, let each try to acquire the Semaphore
    with ThreadPool(TEST_THREADS) as p:
        results = p.map(
            mixed_acquire,
            [(cpp_semaphore, index) for index in range(TEST_THREADS)]
        )
    # Only some of them should succeed
    assert sum(results) == TEST_SEMAPHORE_MAX_COUNT


def test_multiprocess_mixed_with_ctypes_semaphore(ctypes_semaphore: str):
    # Create subprocesses, let each try to acquire the Semaphore
    with ThreadPool(TEST_THREADS) as p:
        results = p.map(
            mixed_acquire,
            [(ctypes_semaphore, index) for index in range(TEST_THREADS)]
        )
    # Only some of them should succeed
    assert sum(results) == TEST_SEMAPHORE_MAX_COUNT


def timed_acquire(name: str) -> float:
    start_time = datetime.datetime.now()
    with OpenSemaphore(name=name) as sem:
        with AcquireSemaphore(sem, timeout_ms=None):
            time.sleep(TEST_SEMAPHORE_ACQUIRE_HOLD_TIME_S)
    end_time = datetime.datetime.now()
    return (end_time - start_time).total_seconds()


def test_multiprocess_infinite_wait():
    # Use two semaphores, let the first lock out the second
    name = unique_name()
    with CreateSemaphore(name):
        with ThreadPool(2) as p:
            results = p.map(
                timed_acquire,
                [name, name]
            )
        fastest = min(results)
        slowest = max(results)
        assert TEST_SEMAPHORE_ACQUIRE_HOLD_TIME_S \
               <= fastest \
               <= (TEST_SEMAPHORE_ACQUIRE_HOLD_TIME_S+50)
        assert (2*TEST_SEMAPHORE_ACQUIRE_HOLD_TIME_S) \
               <= slowest \
               <= (2*TEST_SEMAPHORE_ACQUIRE_HOLD_TIME_S+50)


tests/test_semaphore_win_ctypes.py
"""Tests for `semaphore_win_ctypes` package."""

import datetime
import pytest
import uuid

from ctypes.wintypes import DWORD
from semaphore_win_ctypes import Semaphore, \
    SemaphoreWaitTimeoutException, CreateSemaphore, OpenSemaphore, \
    AcquireSemaphore


@pytest.fixture
def unique_name():
    return str(uuid.uuid4())


def test_basic_with_statements(unique_name):
    with pytest.raises(OSError):
        # Error: Doesn't exist yet
        with OpenSemaphore(unique_name):
            pass

    with CreateSemaphore(unique_name) as created:
        assert created.getvalue() == 1
        with OpenSemaphore(unique_name) as opened:
            assert opened.getvalue() == 1
            with AcquireSemaphore(created, timeout_ms=0) as acquired:
                assert acquired.getvalue() == 0
                with pytest.raises(SemaphoreWaitTimeoutException):
                    # Already acquired
                    with AcquireSemaphore(opened, timeout_ms=0):
                        pass
            with AcquireSemaphore(opened, timeout_ms=0):
                with pytest.raises(SemaphoreWaitTimeoutException):
                    # Already acquired
                    with AcquireSemaphore(created, timeout_ms=0):
                        pass

    with pytest.raises(OSError):
        # Error: CreateSemaphore with statement destroyed the semaphore already
        with OpenSemaphore(unique_name):
            pass


def test_release(unique_name):
    sem = Semaphore(unique_name).create()
    try:
        assert sem.name == unique_name
        sem.acquire(0)
        old_count = sem.release()
        assert old_count == 0
    finally:
        sem.close()


def test_multi_release(unique_name):
    sem = Semaphore(unique_name).create(maximum_count=5, initial_count=5)
    try:
        for _ in range(5):
            sem.acquire(0)
        with pytest.raises(SemaphoreWaitTimeoutException):
            # Can't acquire 6 times
            sem.acquire(0)

        old_count = sem.release(release_count=5)
        assert old_count == 0

        for _ in range(5):
            sem.acquire(0)
        with pytest.raises(SemaphoreWaitTimeoutException):
            # Can't acquire 6 times
            sem.acquire(0)
    finally:
        sem.close()


def test_old_count(unique_name):
    sem = Semaphore(unique_name).create(maximum_count=5, initial_count=5)
    try:
        for _ in range(5):
            sem.acquire(0)
        with pytest.raises(SemaphoreWaitTimeoutException):
            # Can't acquire 6 times
            sem.acquire(0)

        for expected_count in range(5):
            old_count = sem.release(release_count=1)
            assert old_count == expected_count
    finally:
        sem.close()


def test_create_error():
    with pytest.raises(OSError):
        # initial count can't be larger than maximum
        Semaphore().create(maximum_count=1, initial_count=2)


def test_open_without_create(unique_name):
    with pytest.raises(OSError) as _:
        # Doesn't exist, can't open
        Semaphore(unique_name).open()

    # Create it, then it will open
    sem1 = Semaphore(unique_name).create()
    sem2 = Semaphore(unique_name).open()
    assert sem1.getvalue() == 1
    assert sem2.getvalue() == 1
    sem2.close()
    sem1.close()

    with pytest.raises(OSError) as _:
        # No longer exists, can't open
        Semaphore(unique_name).open()


def test_too_many_acquires():
    sem = Semaphore().create()
    try:
        sem.acquire(0)
        with pytest.raises(SemaphoreWaitTimeoutException) as _:
            sem.acquire(0)
    finally:
        sem.close()


def test_too_many_releases():
    sem = Semaphore().create()
    try:
        with pytest.raises(OSError) as _:
            sem.release(release_count=1)
    finally:
        sem.close()


def test_acquire_release_cycle():
    sem = Semaphore().create()
    try:
        for _ in range(10):
            sem.acquire(0)
            old_count = sem.release(release_count=1)
            assert old_count == 0
    finally:
        sem.close()


def test_too_many_closes():
    sem = Semaphore().create()
    sem.close()
    with pytest.raises(OSError) as _:
        sem.close()


def test_low_starting_value():
    sem = Semaphore().create(5, 0)
    try:
        assert sem.getvalue() == 0
        # fail: count is zero
        with pytest.raises(SemaphoreWaitTimeoutException) as _:
            sem.acquire(0)
        # release up to the max of 5
        for expected_count in range(5):
            assert sem.getvalue() == expected_count
            old_count = sem.release(release_count=1)
            assert old_count == expected_count
        # additional release is an error
        with pytest.raises(OSError) as _:
            sem.release(release_count=1)
    finally:
        sem.close()


def test_default_initial_count(unique_name):
    sem = Semaphore(unique_name).create(5)
    try:
        other_sems = []
        other_handles = []
        for _ in range(5):
            other = Semaphore(unique_name)
            handle = other.open()
            handle.acquire(0)
            other_sems.append(other)
            other_handles.append(handle)
        fail_sem = Semaphore(unique_name)
        fail_handle = fail_sem.open()
        with pytest.raises(SemaphoreWaitTimeoutException) as _:
            fail_handle.acquire(0)
        fail_sem.close()
        for other in other_sems:
            other.close()
    finally:
        sem.close()


@pytest.mark.parametrize('timeout_ms', [0, 1, 500])
def test_timeout(timeout_ms: int):
    sem = Semaphore().create()
    try:
        sem.acquire(0)
        start_time = datetime.datetime.now()
        with pytest.raises(SemaphoreWaitTimeoutException) as _:
            sem.acquire(timeout_ms=timeout_ms)
        end_time = datetime.datetime.now()
        delta_ms = (end_time - start_time).total_seconds() * 1000
        assert timeout_ms <= delta_ms <= (timeout_ms+50)
    finally:
        sem.close()


def test_wait_failure():
    sem = Semaphore().create(desired_access=DWORD(0))
    with pytest.raises(OSError):
        sem.acquire()


semaphore_win_ctypes/semaphore_win_ctypes.py
"""Main module."""

from __future__ import annotations
from ctypes import POINTER, windll, WinError
from ctypes.wintypes import BOOL, DWORD, HANDLE, LONG, LPCWSTR, LPVOID
from typing import Union

# https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject
WAIT_OBJECT_0 = 0x00000000
WAIT_ABANDONED = 0x00000080
WAIT_TIMEOUT = 0x00000102
WAIT_FAILED = 0xFFFFFFFF
INFINITE = 0xFFFFFFFF

# https://docs.microsoft.com/en-us/windows/win32/sync/synchronization-object-security-and-access-rights
SEMAPHORE_ALL_ACCESS = 0x1F0003


class SemaphoreWaitTimeoutException(Exception):
    """
    WAIT_TIMEOUT
    """
    pass


LPSECURITY_ATTRIBUTES = LPVOID
LPLONG = POINTER(LONG)

"""
https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createsemaphoreexw
HANDLE CreateSemaphoreExW(
  LPSECURITY_ATTRIBUTES lpSemaphoreAttributes,
  LONG                  lInitialCount,
  LONG                  lMaximumCount,
  LPCWSTR               lpName,
  DWORD                 dwFlags,
  DWORD                 dwDesiredAccess
);
"""
CreateSemaphoreExW = windll.kernel32.CreateSemaphoreExW
CreateSemaphoreExW.argtypes = (LPSECURITY_ATTRIBUTES, LONG, LONG,
                               LPCWSTR, DWORD, DWORD)
CreateSemaphoreExW.restype = HANDLE

"""
https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-opensemaphorew
HANDLE OpenSemaphoreW(
  DWORD   dwDesiredAccess,
  BOOL    bInheritHandle,
  LPCWSTR lpName
);
"""
OpenSemaphoreW = windll.kernel32.OpenSemaphoreW
OpenSemaphoreW.argtypes = DWORD, BOOL, LPCWSTR
OpenSemaphoreW.restype = HANDLE

"""
https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject
DWORD WaitForSingleObject(
  HANDLE hHandle,
  DWORD  dwMilliseconds
);
"""
WaitForSingleObject = windll.kernel32.WaitForSingleObject
WaitForSingleObject.argtypes = HANDLE, DWORD
WaitForSingleObject.restype = DWORD

"""
https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-releasesemaphore
BOOL ReleaseSemaphore(
  HANDLE hSemaphore,
  LONG   lReleaseCount,
  LPLONG lpPreviousCount
);
"""
ReleaseSemaphore = windll.kernel32.ReleaseSemaphore
ReleaseSemaphore.argtypes = HANDLE, LONG, LPLONG
ReleaseSemaphore.restype = BOOL

"""
https://docs.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-closehandle
BOOL CloseHandle(
  HANDLE hObject
);
"""
CloseHandle = windll.kernel32.CloseHandle
CloseHandle.argtypes = (HANDLE,)
CloseHandle.restype = BOOL


class Semaphore:
    def __init__(self, name: str = None):
        """
        Initialize Semaphore class

        :param name: A name for the Semaphore (default: unnamed)
        """
        self.name: str = name
        self.hHandle: HANDLE = HANDLE()

    def create(self,
               maximum_count: int = 1,
               initial_count: int = None,
               desired_access: DWORD = SEMAPHORE_ALL_ACCESS,
               ) -> Semaphore:
        """
        CreateSemaphoreExW

        :param maximum_count: The maximum count of the Semaphore
            (default: 1)
        :param initial_count: The initial count of the Semaphore
            (default: maximum_count)
        :param desired_access: The access mask for the semaphore object
            (default: SEMAPHORE_ALL_ACCESS)
        :raises OSError: The function has failed.
        :returns: The Semaphore, for chaining calls

        https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createsemaphoreexw
        """
        assert not self.hHandle
        if initial_count is None:
            initial_count = maximum_count
        self.hHandle: HANDLE = CreateSemaphoreExW(
            None,
            LONG(initial_count),
            LONG(maximum_count),
            LPCWSTR(self.name),
            DWORD(0),  # reserved
            desired_access,
        )
        if not self.hHandle:
            raise WinError()
        return self

    def open(self,
             desired_access: DWORD = SEMAPHORE_ALL_ACCESS,
             inherit: bool = True,
             ) -> Semaphore:
        """
        OpenSemaphoreW

        :param desired_access: The access mask for the semaphore object
            (default: SEMAPHORE_ALL_ACCESS)
        :param inherit: If this value is TRUE, processes created by this
            process will inherit the handle.
        :raises OSError: The function has failed.
        :returns: The Semaphore, for chaining calls

        https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-opensemaphorew
        """
        assert not self.hHandle
        assert self.name is not None
        self.hHandle: HANDLE = OpenSemaphoreW(
            desired_access,
            BOOL(inherit),
            LPCWSTR(self.name)
        )
        if not self.hHandle:
            raise WinError()
        return self

    def acquire(self, timeout_ms: int = None) -> Semaphore:
        """
        WaitForSingleObject
        :param timeout_ms: The time-out interval, in milliseconds. (default:
            None - infinite wait)
        :raises SemaphoreWaitTimeoutException: The time-out interval elapsed,
            and the object's state is nonsignaled.
        :raises OSError: The function has failed.
        :returns: The Semaphore, for chaining calls

        https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject
        """
        if timeout_ms is None:
            timeout_ms = INFINITE
        else:
            timeout_ms = DWORD(timeout_ms)
            assert timeout_ms != INFINITE, \
                "Use None to specify an infinite timeout"
        ret: DWORD = WaitForSingleObject(
            self.hHandle,
            timeout_ms
        )
        if ret == WAIT_OBJECT_0:
            return self
        elif ret == WAIT_TIMEOUT:
            raise SemaphoreWaitTimeoutException()
        elif ret == WAIT_FAILED:
            raise WinError()
        else:
            assert False, f"Unexpected return code: {ret}"

    def release(self, release_count: int = 1) -> int:
        """
        ReleaseSemaphore
        :param release_count: The amount to increase the semaphore's counter
        :returns: The previous count
        :raises OSError: When release() fails.

        https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-releasesemaphore
        """
        previous_count: LONG = LONG(0)
        ret: BOOL = ReleaseSemaphore(
            self.hHandle,
            LONG(release_count),
            LPLONG(previous_count)
        )
        if not ret:
            raise WinError()
        return previous_count.value

    def close(self) -> None:
        """
        CloseHandle
        :raises OSError: When close() fails.

        https://docs.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-closehandle
        """
        ret: BOOL = CloseHandle(
            self.hHandle
        )
        if not ret:
            raise WinError()
        self.hHandle = None

    def getvalue(self) -> int:
        assert self.hHandle is not None
        try:
            # quickly acquire and release to get the count
            self.acquire(0)
            return self.release() + 1
        except SemaphoreWaitTimeoutException:
            # acquire didn't immediately succeed, count must have been zero
            return 0


class CreateSemaphore:
    def __init__(self,
                 name: str = None,
                 maximum_count: int = 1,
                 initial_count: int = None,
                 desired_access: DWORD = SEMAPHORE_ALL_ACCESS,
                 ):
        self.sem = Semaphore(name)
        self.sem.create(maximum_count, initial_count, desired_access)

    def __enter__(self) -> CreateSemaphore:
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.sem.close()

    def getvalue(self) -> int:
        return self.sem.getvalue()


class OpenSemaphore:
    def __init__(self,
                 name: str = None,
                 desired_access: DWORD = SEMAPHORE_ALL_ACCESS,
                 inherit: bool = True,
                 ):
        self.sem = Semaphore(name)
        self.sem.open(desired_access, inherit)

    def __enter__(self) -> OpenSemaphore:
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.sem.close()

    def getvalue(self) -> int:
        return self.sem.getvalue()


class AcquireSemaphore:
    def __init__(self,
                 handle: Union[CreateSemaphore, OpenSemaphore],
                 timeout_ms: int = None
                 ):
        self.handle = handle
        self.timeout_ms = timeout_ms

    def __enter__(self) -> AcquireSemaphore:
        self.handle.sem.acquire(self.timeout_ms)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.handle.sem.release()

    def getvalue(self) -> int:
        return self.handle.getvalue()


semaphore_win_ctypes/__init__.py
"""Top-level package for Windows Semaphore ctypes."""
from __future__ import annotations
from ctypes import POINTER, windll, WinError
from ctypes.wintypes import BOOL, DWORD, HANDLE, LONG, LPCWSTR, LPVOID
from typing import Union

__author__ = """Robert Alexander"""
__email__ = 'raalexander.phi@gmail.com'
__version__ = '0.1.2'

# https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject
WAIT_OBJECT_0 = 0x00000000
WAIT_ABANDONED = 0x00000080
WAIT_TIMEOUT = 0x00000102
WAIT_FAILED = 0xFFFFFFFF
INFINITE = 0xFFFFFFFF

# https://docs.microsoft.com/en-us/windows/win32/sync/synchronization-object-security-and-access-rights
SEMAPHORE_ALL_ACCESS = 0x1F0003


class SemaphoreWaitTimeoutException(Exception):
    """
    WAIT_TIMEOUT
    """
    pass


LPSECURITY_ATTRIBUTES = LPVOID
LPLONG = POINTER(LONG)

"""
https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createsemaphoreexw
HANDLE CreateSemaphoreExW(
  LPSECURITY_ATTRIBUTES lpSemaphoreAttributes,
  LONG                  lInitialCount,
  LONG                  lMaximumCount,
  LPCWSTR               lpName,
  DWORD                 dwFlags,
  DWORD                 dwDesiredAccess
);
"""
CreateSemaphoreExW = windll.kernel32.CreateSemaphoreExW
CreateSemaphoreExW.argtypes = (LPSECURITY_ATTRIBUTES, LONG, LONG,
                               LPCWSTR, DWORD, DWORD)
CreateSemaphoreExW.restype = HANDLE

"""
https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-opensemaphorew
HANDLE OpenSemaphoreW(
  DWORD   dwDesiredAccess,
  BOOL    bInheritHandle,
  LPCWSTR lpName
);
"""
OpenSemaphoreW = windll.kernel32.OpenSemaphoreW
OpenSemaphoreW.argtypes = DWORD, BOOL, LPCWSTR
OpenSemaphoreW.restype = HANDLE

"""
https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject
DWORD WaitForSingleObject(
  HANDLE hHandle,
  DWORD  dwMilliseconds
);
"""
WaitForSingleObject = windll.kernel32.WaitForSingleObject
WaitForSingleObject.argtypes = HANDLE, DWORD
WaitForSingleObject.restype = DWORD

"""
https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-releasesemaphore
BOOL ReleaseSemaphore(
  HANDLE hSemaphore,
  LONG   lReleaseCount,
  LPLONG lpPreviousCount
);
"""
ReleaseSemaphore = windll.kernel32.ReleaseSemaphore
ReleaseSemaphore.argtypes = HANDLE, LONG, LPLONG
ReleaseSemaphore.restype = BOOL

"""
https://docs.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-closehandle
BOOL CloseHandle(
  HANDLE hObject
);
"""
CloseHandle = windll.kernel32.CloseHandle
CloseHandle.argtypes = (HANDLE,)
CloseHandle.restype = BOOL


class Semaphore:
    def __init__(self, name: str = None):
        """
        Initialize Semaphore class

        :param name: A name for the Semaphore (default: unnamed)
        """
        self.name: str = name
        self.hHandle: HANDLE = HANDLE()

    def create(self,
               maximum_count: int = 1,
               initial_count: int = None,
               desired_access: DWORD = SEMAPHORE_ALL_ACCESS,
               ) -> Semaphore:
        """
        CreateSemaphoreExW

        :param maximum_count: The maximum count of the Semaphore
            (default: 1)
        :param initial_count: The initial count of the Semaphore
            (default: maximum_count)
        :param desired_access: The access mask for the semaphore object
            (default: SEMAPHORE_ALL_ACCESS)
        :raises OSError: The function has failed.
        :returns: The Semaphore, for chaining calls

        https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createsemaphoreexw
        """
        assert not self.hHandle
        if initial_count is None:
            initial_count = maximum_count
        self.hHandle: HANDLE = CreateSemaphoreExW(
            None,
            LONG(initial_count),
            LONG(maximum_count),
            LPCWSTR(self.name),
            DWORD(0),  # reserved
            desired_access,
        )
        if not self.hHandle:
            raise WinError()
        return self

    def open(self,
             desired_access: DWORD = SEMAPHORE_ALL_ACCESS,
             inherit: bool = True,
             ) -> Semaphore:
        """
        OpenSemaphoreW

        :param desired_access: The access mask for the semaphore object
            (default: SEMAPHORE_ALL_ACCESS)
        :param inherit: If this value is TRUE, processes created by this
            process will inherit the handle.
        :raises OSError: The function has failed.
        :returns: The Semaphore, for chaining calls

        https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-opensemaphorew
        """
        assert not self.hHandle
        assert self.name is not None
        self.hHandle: HANDLE = OpenSemaphoreW(
            desired_access,
            BOOL(inherit),
            LPCWSTR(self.name)
        )
        if not self.hHandle:
            raise WinError()
        return self

    def acquire(self, timeout_ms: int = None) -> Semaphore:
        """
        WaitForSingleObject
        :param timeout_ms: The time-out interval, in milliseconds. (default:
            None - infinite wait)
        :raises SemaphoreWaitTimeoutException: The time-out interval elapsed,
            and the object's state is nonsignaled.
        :raises OSError: The function has failed.
        :returns: The Semaphore, for chaining calls

        https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject
        """
        if timeout_ms is None:
            timeout_ms = INFINITE
        else:
            timeout_ms = DWORD(timeout_ms)
            assert timeout_ms != INFINITE, \
                "Use None to specify an infinite timeout"
        ret: DWORD = WaitForSingleObject(
            self.hHandle,
            timeout_ms
        )
        if ret == WAIT_OBJECT_0:
            return self
        elif ret == WAIT_TIMEOUT:
            raise SemaphoreWaitTimeoutException()
        elif ret == WAIT_FAILED:
            raise WinError()
        else:
            assert False, f"Unexpected return code: {ret}"

    def release(self, release_count: int = 1) -> int:
        """
        ReleaseSemaphore
        :param release_count: The amount to increase the semaphore's counter
        :returns: The previous count
        :raises OSError: When release() fails.

        https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-releasesemaphore
        """
        previous_count: LONG = LONG(0)
        ret: BOOL = ReleaseSemaphore(
            self.hHandle,
            LONG(release_count),
            LPLONG(previous_count)
        )
        if not ret:
            raise WinError()
        return previous_count.value

    def close(self) -> None:
        """
        CloseHandle
        :raises OSError: When close() fails.

        https://docs.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-closehandle
        """
        ret: BOOL = CloseHandle(
            self.hHandle
        )
        if not ret:
            raise WinError()
        self.hHandle = None

    def getvalue(self) -> int:
        assert self.hHandle is not None
        try:
            # quickly acquire and release to get the count
            self.acquire(0)
            return self.release() + 1
        except SemaphoreWaitTimeoutException:
            # acquire didn't immediately succeed, count must have been zero
            return 0


class CreateSemaphore:
    def __init__(self,
                 name: str = None,
                 maximum_count: int = 1,
                 initial_count: int = None,
                 desired_access: DWORD = SEMAPHORE_ALL_ACCESS,
                 ):
        self.sem = Semaphore(name)
        self.sem.create(maximum_count, initial_count, desired_access)

    def __enter__(self) -> CreateSemaphore:
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.sem.close()

    def getvalue(self) -> int:
        return self.sem.getvalue()


class OpenSemaphore:
    def __init__(self,
                 name: str = None,
                 desired_access: DWORD = SEMAPHORE_ALL_ACCESS,
                 inherit: bool = True,
                 ):
        self.sem = Semaphore(name)
        self.sem.open(desired_access, inherit)

    def __enter__(self) -> OpenSemaphore:
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.sem.close()

    def getvalue(self) -> int:
        return self.sem.getvalue()


class AcquireSemaphore:
    def __init__(self,
                 handle: Union[CreateSemaphore, OpenSemaphore],
                 timeout_ms: int = None
                 ):
        self.handle = handle
        self.timeout_ms = timeout_ms

    def __enter__(self) -> AcquireSemaphore:
        self.handle.sem.acquire(self.timeout_ms)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.handle.sem.release()

    def getvalue(self) -> int:
        return self.handle.getvalue()
